test/integration/healthcheck/healthcheck_operation.go

```diff
@@ -0,0 +1,185 @@
+		Name:      extensionName,
+	}, healthConditionType, gardencorev1beta1.ConditionFalse, healthcheck.HealthCheckUnsuccessful)
+}
+
+func ControlPlaneHealthCheckDeleteSeedDeployment(ctx context.Context, f *framework.ShootFramework, controlPlaneName, deploymentName string, healthConditionType gardencorev1beta1.ConditionType) error {
+	return deleteSeedDeploymentCheck(ctx, f, extensionsv1alpha1.SchemeGroupVersion.WithKind(extensionsv1alpha1.ControlPlaneResource), controlPlaneName, deploymentName, healthConditionType)
+}
+
+func WorkerHealthCheckDeleteSeedDeployment(ctx context.Context, f *framework.ShootFramework, controlPlaneName, deploymentName string, healthConditionType gardencorev1beta1.ConditionType) error {
+	return deleteSeedDeploymentCheck(ctx, f, extensionsv1alpha1.SchemeGroupVersion.WithKind(extensionsv1alpha1.WorkerResource), controlPlaneName, deploymentName, healthConditionType)
+}
+
+func deleteSeedDeploymentCheck(ctx context.Context, f *framework.ShootFramework, extensionKind schema.GroupVersionKind, controlPlaneName, deploymentName string, healthConditionType gardencorev1beta1.ConditionType) error {
+	var err error
+
+	cloudControllerDeployment := appsv1.Deployment{ObjectMeta: metav1.ObjectMeta{Name: deploymentName, Namespace: f.ShootSeedNamespace()}}
+	if err := f.SeedClient.Client().Delete(ctx, &cloudControllerDeployment); err != nil {
+		return err
+	}
+	defer func() {
+		err = f.GardenerFramework.UpdateShoot(ctx, f.Shoot, func(shoot *gardencorev1beta1.Shoot) error {
+			shoot.Annotations[common.ShootOperation] = common.ShootOperationReconcile
+			return nil
+		})
+		framework.ExpectNoError(err)
+
+		// then make sure the condition is fine again after reconciliation
+		err = operation.WaitForExtensionCondition(ctx, f.Logger, f.SeedClient.Client(), extensionKind, types.NamespacedName{
+			Namespace: f.ShootSeedNamespace(),
+			Name:      controlPlaneName,
+		}, healthConditionType, gardencorev1beta1.ConditionTrue, healthcheck.HealthCheckSuccessful)
+		framework.ExpectNoError(err)
+	}()
+	return operation.WaitForExtensionCondition(ctx, f.Logger, f.SeedClient.Client(), extensionKind, types.NamespacedName{
+		Namespace: f.ShootSeedNamespace(),
+		Name:      controlPlaneName,
+	}, healthConditionType, gardencorev1beta1.ConditionUnknown, gardencorev1beta1.ConditionCheckError)
+}
+
+func MachineDeletionHealthCheck(ctx context.Context, f *framework.ShootFramework) error {
+	var err error
+	machineList := machinev1alpha1.MachineList{}
+	if err := f.SeedClient.Client().List(ctx, &machineList, client.InNamespace(f.ShootSeedNamespace())); err != nil {
+		return err
+	}
+
+	if len(machineList.Items) == 0 {
+		return fmt.Errorf("trying to delete machine as part of health check test from a cluster with no nodes (seed: %s, namespace: %s)", f.Seed.Name, f.ShootSeedNamespace())
+	}
+
+	machine := machineList.Items[0]
+	err = f.SeedClient.Client().Delete(ctx, &machine)
+	if err != nil {
+		return err
+	}
+	defer func() {
```

Comment:
the error handling is mixed between returning an error and panic on error (ginkgo Expect no error).
I think we should stay with one implementation so returning an error at any time? wdyt?

Reply:
The reason is, that the cleanup functionality is in the defer function to make sur eit is being executed.   I could either return a []error that includes potential cleanup errors, or let it ginko fail. I think it is appropriate to let the test suite fail of the cleanup is not successful. IDK what kind of error handling the caller of this function could do if the cleanup failed / the test def. should not continue.

Reply:
If I understand the test correctly:
The machine is deleted and the test waits for the worker extensions to be unhealthy.
Then the defer function will be called that validates that the mcm correctly reconciles the deleted machine and the worker extensions gets healthy again.

Wouldn't it be sufficient to check the successful healthiness after the unsuccessful. 
Why do we want to check if the worker is healthy when the unhealthy checks fails?

So just:
```go

err := operation.WaitForExtensionCondition(ctx, f.Logger, f.SeedClient.Client(), extensionsv1alpha1.SchemeGroupVersion.WithKind(extensionsv1alpha1.WorkerResource), types.NamespacedName{
	Namespace: f.ShootSeedNamespace(),
	Name:      f.Shoot.GetName(),
}, gardencorev1beta1.ShootEveryNodeReady, gardencorev1beta1.ConditionFalse, healthcheck.HealthCheckUnsuccessful)
framework.ExpectNoError(err) // alternatively return err

err := operation.WaitForExtensionCondition(ctx, f.Logger, f.SeedClient.Client(), extensionsv1alpha1.SchemeGroupVersion.WithKind(extensionsv1alpha1.WorkerResource), types.NamespacedName{
	Namespace: f.ShootSeedNamespace(),
	Name:      f.Shoot.GetName(),
}, gardencorev1beta1.ShootEveryNodeReady, gardencorev1beta1.ConditionTrue, healthcheck.HealthCheckSuccessful)
framework.ExpectNoError(err) // alternatively return err
```

Reply:
Thanks!